File: api/views/user_weather_alerts.py
import json
from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt
from django.contrib.gis.geos import Point
from ..models import UserWeatherAlert

@csrf_exempt
def user_weather_alerts(request):
    """
    GET: Retrieves all user weather alerts
    POST: Creates a new user weather alert
    """
    if request.method == "GET":
        alerts = UserWeatherAlert.objects.filter(active=True).values(
            'id', 'name', 'condition', 'threshold', 'comparison', 'active'
        )
        return JsonResponse(list(alerts), safe=False)
    
    elif request.method == "POST":
        try:
            data = json.loads(request.body)
            name = data.get('name')
            condition = data.get('condition')
            threshold = data.get('threshold')
            comparison = data.get('comparison')
            
            if not all([name, condition, threshold, comparison]):
                return JsonResponse({"error": "Missing required fields"}, status=400)
            
            lat = data.get('latitude')
            lon = data.get('longitude')
            location = Point(float(lon), float(lat)) if lat and lon else None
            
            alert = UserWeatherAlert.objects.create(
                name=name,
                condition=condition,
                threshold=float(threshold),
                comparison=comparison,
                location=location
            )
            
            return JsonResponse({
                "id": alert.id,
                "name": alert.name,
                "condition": alert.condition,
                "threshold": alert.threshold,
                "comparison": alert.comparison,
                "active": alert.active
            }, status=201)
            
        except Exception as e:
            return JsonResponse({"error": str(e)}, status=400)

@csrf_exempt
def user_weather_alert_detail(request, alert_id):
    """
    GET: Retrieves a specific user weather alert
    PUT: Updates a specific user weather alert
    DELETE: Deletes a specific user weather alert
    """
    try:
        alert = UserWeatherAlert.objects.get(id=alert_id)
    except UserWeatherAlert.DoesNotExist:
        return JsonResponse({"error": "Alert not found"}, status=404)
    
    if request.method == "GET":
        return JsonResponse({
            "id": alert.id,
            "name": alert.name,
            "condition": alert.condition,
            "threshold": alert.threshold,
            "comparison": alert.comparison,
            "active": alert.active
        })
    
    elif request.method == "PUT":
        try:
            data = json.loads(request.body)
            if 'name' in data:
                alert.name = data['name']
            if 'condition' in data:
                alert.condition = data['condition']
            if 'threshold' in data:
                alert.threshold = float(data['threshold'])
            if 'comparison' in data:
                alert.comparison = data['comparison']
            if 'active' in data:
                alert.active = data['active']
            
            alert.save()
            return JsonResponse({
                "id": alert.id,
                "name": alert.name,
                "condition": alert.condition,
                "threshold": alert.threshold,
                "comparison": alert.comparison,
                "active": alert.active
            })
            
        except Exception as e:
            return JsonResponse({"error": str(e)}, status=400)
    
    elif request.method == "DELETE":
        alert.delete()
        return JsonResponse({"message": "Alert deleted successfully"})

@csrf_exempt
def check_user_alerts(request):
    """
    Checks if current weather conditions match any user alerts
    """
    if request.method == "POST":
        try:
            data = json.loads(request.body)
            weather_data = data.get('weather_data')
            
            if not weather_data:
                return JsonResponse({"error": "No weather data provided"}, status=400)
            
            # Get all active alerts
            alerts = UserWeatherAlert.objects.filter(active=True)
            matched_alerts = []
            
            for alert in alerts:
                if check_alert_condition(alert, weather_data):
                    matched_alerts.append({
                        "id": alert.id,
                        "name": alert.name,
                        "message": f"{alert.name}: {alert.get_condition_display()} is {alert.get_comparison_display()} {alert.threshold}"
                    })
            
            return JsonResponse({"matched_alerts": matched_alerts})
            
        except Exception as e:
            return JsonResponse({"error": str(e)}, status=400)

def check_alert_condition(alert, weather_data):
    """Helper function to check if weather data matches alert condition"""
    # Get the first weather data point (current conditions)
    current = weather_data[0] if isinstance(weather_data, list) else weather_data
    
    value = None
    if alert.condition == 'SUNNY':
        value = 100 - current.get('cloudiness', 0)  # Sunny is inverse of cloudiness
    elif alert.condition == 'RAINY':
        value = current.get('rain', 0)
    elif alert.condition == 'WINDY':
        value = current.get('wind_speed', 0)
    elif alert.condition == 'HOT' or alert.condition == 'COLD':
        value = current.get('temperature', 0)
    
    if value is None:
        return False
    
    if alert.comparison == 'GT':
        return value > alert.threshold
    elif alert.comparison == 'LT':
        return value < alert.threshold
    elif alert.comparison == 'EQ':
        return abs(value - alert.threshold) < 0.1  # Approximate equality
    
    return False 

File: client/src/components/UserWeatherAlerts.js
import React, { useState, useEffect } from 'react';
import './UserWeatherAlerts.css';

const UserWeatherAlerts = ({ weatherData, latitude, longitude }) => {
  const BASE_URL = process.env.REACT_APP_API_URL;
  const [alerts, setAlerts] = useState([]);
  const [matchedAlerts, setMatchedAlerts] = useState([]);
  const [showForm, setShowForm] = useState(false);
  const [newAlert, setNewAlert] = useState({
    name: '',
    condition: 'SUNNY',
    threshold: 70,
    comparison: 'GT'
  });
  const [error, setError] = useState('');
  const [success, setSuccess] = useState('');

  const fetchUserAlerts = async () => {
    try {
      setError('');
      const response = await fetch(`${BASE_URL}/api/user-weather-alerts/`);
      if (!response.ok) {
        if (response.status === 500) {
          throw new Error("Database migration needed. Please contact an administrator.");
        }
        throw new Error("Failed to fetch user alerts");
      }
      const data = await response.json();
      setAlerts(data);
    } catch (err) {
      console.error("Error fetching alerts:", err);
      setError("Error fetching user alerts: " + err.message);
    }
  };

  const checkAlertConditions = async () => {
    if (!weatherData || !weatherData.length) return;

    try {
      const response = await fetch(`${BASE_URL}/api/user-weather-alerts/check/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ weather_data: weatherData }),
      });

      if (!response.ok) {
        throw new Error("Failed to check alerts");
      }
      
      const data = await response.json();
      setMatchedAlerts(data.matched_alerts || []);
    } catch (err) {
      setError("Error checking alert conditions: " + err.message);
    }
  };

  useEffect(() => {
    fetchUserAlerts();
  }, []);

  useEffect(() => {
    if (weatherData && weatherData.length > 0) {
      checkAlertConditions();
    }
  }, [weatherData]);

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setNewAlert({
      ...newAlert,
      [name]: value
    });
  };

  const handleCreateAlert = async (e) => {
    e.preventDefault();
    
    try {
      const alertData = {
        ...newAlert,
        latitude: latitude,
        longitude: longitude
      };
      
      const response = await fetch(`${BASE_URL}/api/user-weather-alerts/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(alertData),
      });

      if (!response.ok) {
        throw new Error("Failed to create alert");
      }
      
      setNewAlert({
        name: '',
        condition: 'SUNNY',
        threshold: 70,
        comparison: 'GT'
      });
      setShowForm(false);
      setSuccess("Alert created successfully!");
      
      setTimeout(() => setSuccess(''), 3000);
      
      fetchUserAlerts();
    } catch (err) {
      setError("Error creating alert: " + err.message);
    }
  };

  const handleDeleteAlert = async (id) => {
    try {
      const response = await fetch(`${BASE_URL}/api/user-weather-alerts/${id}/`, {
        method: 'DELETE',
      });

      if (!response.ok) {
        throw new Error("Failed to delete alert");
      }
      
      fetchUserAlerts();
      setSuccess("Alert deleted successfully!");
      
      setTimeout(() => setSuccess(''), 3000);
    } catch (err) {
      setError("Error deleting alert: " + err.message);
    }
  };

  const getConditionText = (condition) => {
    const conditions = {
      'SUNNY': 'Sunny (Low cloudiness)',
      'RAINY': 'Rainy (Precipitation)',
      'WINDY': 'Windy (High wind speed)',
      'HOT': 'Hot temperature',
      'COLD': 'Cold temperature'
    };
    return conditions[condition] || condition;
  };

  const getComparisonText = (comparison) => {
    const comparisons = {
      'GT': 'greater than',
      'LT': 'less than',
      'EQ': 'equal to'
    };
    return comparisons[comparison] || comparison;
  };

  const getConditionUnit = (condition) => {
    switch(condition) {
      case 'SUNNY':
      case 'RAINY':
        return '%';
      case 'WINDY':
        return 'km/h';
      case 'HOT':
      case 'COLD':
        return '°C';
      default:
        return '';
    }
  };

  return (
    <div className="user-weather-alerts">
      <h3>Custom Weather Alerts</h3>
      
      {}
      {success && (
        <div className="alert-success">
          {success}
        </div>
      )}
      
      {}
      {error && (
        <div className="alert-error">
          {error}
          <button onClick={() => setError('')}>×</button>
        </div>
      )}
      
      {}
      {matchedAlerts.length > 0 && (
        <div className="matched-alerts">
          <h4>Active Alerts</h4>
          {matchedAlerts.map((alert, index) => (
            <div key={index} className="matched-alert">
              <span className="alert-icon">⚠️</span> {alert.message}
            </div>
          ))}
        </div>
      )}
      
      {}
      {showForm ? (
        <div className="alert-form">
          <h4>Create New Alert</h4>
          <form onSubmit={handleCreateAlert}>
            <div className="form-group">
              <label>Name:</label>
              <input 
                type="text"
                name="name"
                value={newAlert.name}
                onChange={handleInputChange}
                placeholder="Alert name"
                required
              />
            </div>
            
            <div className="form-group">
              <label>Condition:</label>
              <select
                name="condition"
                value={newAlert.condition}
                onChange={handleInputChange}
                required
              >
                <option value="SUNNY">Sunny (Low cloudiness)</option>
                <option value="RAINY">Rainy (Precipitation)</option>
                <option value="WINDY">Windy (High wind speed)</option>
                <option value="HOT">Hot temperature</option>
                <option value="COLD">Cold temperature</option>
              </select>
            </div>
            
            <div className="form-group">
              <label>Comparison:</label>
              <select
                name="comparison"
                value={newAlert.comparison}
                onChange={handleInputChange}
                required
              >
                <option value="GT">Greater than</option>
                <option value="LT">Less than</option>
                <option value="EQ">Equal to</option>
              </select>
            </div>
            
            <div className="form-group">
              <label>Threshold ({getConditionUnit(newAlert.condition)}):</label>
              <input 
                type="number"
                name="threshold"
                value={newAlert.threshold}
                onChange={handleInputChange}
                required
              />
            </div>
            
            <div className="form-buttons">
              <button type="submit" className="btn-create">Create Alert</button>
              <button type="button" className="btn-cancel" onClick={() => setShowForm(false)}>Cancel</button>
            </div>
          </form>
        </div>
      ) : (
        <button className="btn-add-alert" onClick={() => setShowForm(true)}>
          + Add Custom Alert
        </button>
      )}
      
      {}
      {alerts.length > 0 && (
        <div className="alerts-list">
          <h4>Your Custom Alerts</h4>
          {alerts.map((alert) => (
            <div key={alert.id} className="user-alert">
              <div className="alert-content">
                <strong>{alert.name}</strong>
                <p>Alert when {getConditionText(alert.condition)} is {getComparisonText(alert.comparison)} {alert.threshold}{getConditionUnit(alert.condition)}</p>
              </div>
              <button className="btn-delete" onClick={() => handleDeleteAlert(alert.id)}>Delete</button>
            </div>
          ))}
        </div>
      )}
    </div>
  );
};

export default UserWeatherAlerts; 